src/io.c

```diff
@@ -1093,6 +1094,7 @@ static void PutLine2(TypOutputFile * output, const Char * line, UInt len)
     /* special handling of string streams, where we can copy directly */
     if (output->isstringstream) {
       str = CONST_ADDR_OBJ(output->stream)[1];
+      ConvString(str);
```

Comment:
This is still unsafe in that `ConvString` doesn't check whether the given object is a string. So at the very least we should insert an assertion here:
```suggestion
      GAP_ASSERT(IS_STRING(str));
      ConvString(str);
```
Of course even better might be to always handle this, but given that we are in the middle of outputting text, just calling `Error` blindly may be a really bad idea... We could also just do `if (!IS_STRING(str)) Panic(...)` (with the idea that crash is better than silently corrupting data). Perhaps somewhat better is to simply fall back to the generic code in this case (but chances are it'll also just end up running into an `Error`. BTW, one can also use `IsStringConv`... so like this:
```
if (output->isstringstream) {
  str = ..
  if (IsStringConv(str)) {
    lstr = ..
    ...
    return;
  }
}
```

Reply:
I think it will get caught in ConvString eventually, but something clearer might be nice. Also, asserts are only in debug builds anyway.

I agree that IsStringConv, then heading back to the GAP level path (which will Error rather than crash) is the best option. Sorry @embray we are pushing this around so much.

Reply:
It is true that in debug mode (with assertion on), `ConvString` will catch bad values (but indirectly: first `LEN_LIST` will assert out if it's not a list, then later `CHAR_VALUE` if its values are not vars.

But in release mode, only `LEN_LIST` will catch non-lists, but `CHAR_VALUE` will produce garbage. And if the input happens to be a list of immediate integers, it might segfault.

But in all of these cases, there won't be a very helpful error message... It might be insightful to try and triggers this on purpose, and see what happens?

Anyway, I do not yet see how any of this could explain the 32bit crash... I will try and see if I can reproduce it on an Ubuntu 20.20 box.

Reply:
Yes, that 32-bit crash I'm unable to reproduce myself (I might have another try)

Reply:
Reading the docs for OutputTextString, it even talks about "puts all received characters into the list 'list'", so a user could reasonably read that, and use OutputTextString in lists which are not strings

```
gap> x := [1];
[ 1 ]
gap> w := OutputTextString(x, false);
OutputTextString(0)
gap> PrintTo(w, "hello");
gap> x;
"hello"
gap> x := [1];
[ 1 ]
gap> w := OutputTextString(x, true);
OutputTextString(1)
gap> PrintTo(w, "hello");
gap> x;
[ 1, 'h', 'e', 'l', 'l', 'o' ]
```

Reply:
evil :-)

Reply:
I've restarted the jobs once again, just rule out a random fluke (which would still be worrisome, but might not be directly caused by this PR).